### ISSUE TITLE
Back-calculating key unit parameters to achieve TEA targets

### ISSUE THREAD

**Issue by yalinli2** [BioSTEAMDevelopmentGroup/biosteam#22]
**Is your feature request related to a problem? Please describe.**
BioSTEAM is great on calculating TEA metrics like stream price/IRR with a given system design, I'm wondering, would it be possible to/do you see value in letting BioSTEAM automate the design so that a certain price/IRR target can be reached?

For example, let's say for the cornstover biorefinery, we assume a 95% glucose-to-ethanol conversion that gives us an MESP of ~$0.69/kg, so can BioSTEAM calculate how low this conversion can be to give an MESP of $1/kg or less?

**Describe the solution you'd like**
The fundamental question I'd like to address is: what would it take to reach a TEA target? I consider this question in three different levels based on what I know about BioSTEAM (easiest to hardest to implement), it's kind of similar to Monte Carlo:
 - Easiest: as I described in the example. We can BioSTEAM calculate what the MESP will be from 0-100% conversion, then select the cutoff conversion to reach the target.

- Medium (this might not be a robust solution): we provide a list of parameters with their probable ranges, BioSTEAM runs sensitivity analysis to rank these parameters from most to least impactful, then BioSTEAM changes the most impactful parameter to see if the target can be reached. If not, BioSTEAM will include the second most impactful parameter, run Monte Carlo within the given ranges, then see if the target can be reached, so on and so forth.

    - _Note, I considered whether it'd be better to let BioSTEAM set the most impactful parameter at the value that would be closest to the target and only manipulate the second most impactful parameter to save some computation power, but realized this might result in local optima_

 - Hardest: we provide a list of parameters with their probable ranges (or even better, BioSTEAM loads some default parameters like the `biosteam.evaluation.State. load_default_parameters` function), and BioSTEAM calculates the probability at a given value with the distribution of a certain parameter (i.e., the cumulative distribution function). Then BioSTEAM will do Monte Carlo for all the parameters we provided and give each simulation scenario the probability score (maybe by multiplying all probabilities?) and rank these scenarios from most to least probable. This way, we can clearly see:
     - How probable to achieve a TEA target?
     - What assumptions we need to make to achieve that target?

Look forward to your thoughts, thanks!!!

**Comment by yoelcortes** [comment]
As for your first analysis, we could use scipy's minimization or root solving methods. They are pretty easy to use. You can pass the bounds for all parameters and solve for the values that give the minimum or the target IRR/Price, or whatever metric you pass to the objective function. Here is an example of how this could be done:

Single variable solution for cut-off conversion
```python
from scipy.optimize import brentq
from biorefineries.cornstover.system import cornstover_sys, cornstover_tea, R301, R302

# Should return 0 where IRR = 0.15
def f(x): # objective function
    R301.saccharification[2].X = x # Set glucan to glucose conversion of R301
    cornstover_sys.simulate()
   return  0.15 - cornstover_tea.solve_IRR()
x = brentq(f, 0, 1) # Done!
```

Multivariate solution to maximize IRR
```python
from scipy.optimize import minimize
from biorefineries.cornstover.model import cornstover_model
def f(x): # objective function
    # p1, p2, p3, ... = x
    # We can set variables anyway we want here.
    # We could also use BioSTEAM's Model object to do this.
    results = cornstover_model(x)
   return - results['IRR'] # We are minimizing negative IRR
solution = minimize(f, bounds, ...)
p1, p2, p3, ... = solution.x
```
I don't believe we should add these methods inside biosteam objects. We would still be requesting the same info (e.g. an objective function, bounds, and all the other optional arguments in scipy functions). However, once someone gets this done, we can add their code as an "example recipe" in the docs.

As for your second analysis (the Medium one). Running Monte Carlo again and again for each parameter we change would take several weeks. Wouldn't all the data required to find this already be in your first Monte Carlo analysis? I would suggest preprocessing the data, then using [sklearn's SVR](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html) to build a machine learning (ML) model to perform the calculations (to vary whatever parameters you need). We can add a method in biosteam to preprocess data and return a ML model, for those people who are timid about ML.

As for the last analysis (the Hardest one). I don't think this is all that hard. You have the distributions of all parameters in the Model object, and each [chaospy Distribution](https://chaospy.readthedocs.io/en/master/distributions/index.html) object have a method to get cumulative distribution. All Monte Carlo results are saved in a data frame, just need to pass these values to each parameter.

**Comment by yalinli2** [comment]
Thanks for the explanations and helpful links, I'll give the last one a try and see how that goes.

My proposed approaches aside, what would you do if you were to solve the question I tried to address - what unit/process assumptions are needed to meet a certain TEA target? Are there simpler or more robust ways? Thanks!

**Comment by yoelcortes** [comment]
Hmm, if what we want is the worst each parameter can perform to meet a certain TEA target. Then, after we know the relationship between all parameters and the target (e.g. through Spearman coefficients), we could brute test a grid of all parameter values (for just the ones that matter the most). You can make a grid of values using numpy, pass it to the model using the load_samples method, evaluate, then filter the scenarios. This should give us all the results we need to conduct further analysis. For example, we can get all scenarios where a parameter is at its minimum

```python
import numpy as np
# Let's say results is an array where each row is a scenario
# and each column is a parameter.
p0s = results[:, 0]
p0_min = results.min() # minimum possible value for parameter 0
index, = np.where(np.p0s ==p0_min)
scenarios_with_p0_min = results[index]
```
This is just one analysis you could do. But the grid should give you all the data you need.

**Comment by yalinli2** [comment]
Yeah I think what you proposed works, I think Monte Carlo analysis is sufficient to answer my question, I just need to process/interpret the data and come up with a way to define the probability of each scenario. Many thanks!!!
